### PR TITLE
Add array and map operators to the compiler.

### PR DESCRIPTION
**Pull request by sellout**
SD-995 #done, SD-996 #done

Unshift and shift of array and map keys and values.

Also rename “object” to “map” in affected places (more renaming to be
done as part of SD-1018).

### CODE REVIEW

**Comment by drostron** [comment]
👍🏽

Couple of items related to my lack of familiarity:
* I would like gain a better understanding on what's missing from mongodb that's necessary to support FlattenArrayIndices, ShiftArray, FlattenMapKeys.
* `foo{"bar"}` doesn't seem to be supported. What am I missing?
* Am I missing map concat?

**Comment by sellout** [comment]
I don’t think anything is missing from Mongo to support these – it’s just not part of this issue. The SQL² epic is intended to implement the full parser/compiler (IE, the frontend) without worrying about implementation for any particular backend.

`foo{bar}` should already be supported as `FieldDeref` / `structural.ObjectProject`, not new in this PR.

We don’t have any kind of concat reduction yet, so I don’t think mapconcat is possible. I could be overlooking something, though.

**Comment by jdegoes** [comment]
Array concat is `||`, inherited from Postgres, map concat could be anything we want. I'm partial to `++`, although you can call the function manually as for all operators (`MAP_CONCAT`?). 

The corresponding reductions have syntax, e.g. `[foo ...]` reduces the last dimensional set into an array, while `{ keys : values ...}` does the same for maps.

**Comment by sellout** [comment]
Ohhhhh, concatting two maps, not `mapconcat`. Got it!

File: core/src/main/scala/quasar/std/structural.scala [hunk 1]
```diff
@@ -220,10 +220,98 @@ trait StructuralLib extends Library {
     },
     untyper(tpe => success(List(FlexArr(0, None, tpe)))))
 
+  val FlattenMapKeys = Expansion(
+    "{*:}",
+    "Flattens a map into a set",
```
Comment (wemrysi) [inline core/src/main/scala/quasar/std/structural.scala]: This just flattens the keys, no? Not the entire map?

File: core/src/main/scala/quasar/std/structural.scala [hunk 2]
```diff
@@ -220,10 +220,98 @@ trait StructuralLib extends Library {
...
+    Set(Top), AnyObject :: Nil,
+    noSimplification,
+    partialTyper { case List(x) if x.objectLike => Str },
+    untyper(tpe => success(List(Obj(Map(), Some(Top))))))
+
+  val FlattenArrayIndices = Expansion(
+    "[*:]",
+    "Flattens an array into a set",
```
Comment (wemrysi) [inline core/src/main/scala/quasar/std/structural.scala]: Same here no, this doesn't flatten the contents, just the indices, right?

File: core/src/main/scala/quasar/std/structural.scala [hunk 3]
```diff
@@ -220,10 +220,98 @@ trait StructuralLib extends Library {
...
+    Set(Int), AnyArray :: Nil,
+    noSimplification,
+    partialTyper { case List(x) if x.arrayLike => Int },
+    partialUntyper {
+      case Set(Int) | Int => List(FlexArr(0, None, Top))
+    })
+
+  val ShiftMap = Expansion(
+    "SHIFT_MAP",
+    "Shifts a map into a set",
```
Comment (wemrysi) [inline core/src/main/scala/quasar/std/structural.scala]: This (and the rest of the help text on these new functions) leaves a bit to be desired, especially if this is ever displayed to a user. I'd prefer something a bit more descriptive, this doesn't really convey any more information than the name of the function itself (the fact that it returns a set, but that's about it).

Comment (sellout) [inline core/src/main/scala/quasar/std/structural.scala]: Got any suggestions as to what to say? If we had some MRA docs, I think referencing the definitions of “shift”, “unshift”, and “flatten” there would be a good idea.

Comment (wemrysi) [inline core/src/main/scala/quasar/std/structural.scala]: Heh, well my comment was prompted by the fact that I don't remember what "shift" and "unshift" mean ;) At the risk of being verbose, would something like "Shifts a map into a set, <explain what that actually means>" work?

Comment (wemrysi) [inline core/src/main/scala/quasar/std/structural.scala]: And I agree regarding referencing some docs, should they exist.

Comment (jdegoes) [inline core/src/main/scala/quasar/std/structural.scala]: 'Zooms into the array elements, looking at the indices,' etc.

**Pull request merged**